arangod/Aql/OptimizerRules.cpp

```diff
@@ -1943,101 +1943,107 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
     // test if we can use an alternative version of COLLECT with a hash table
     bool const canUseHashAggregation =
         (!groupVariables.empty() && collectNode->getOptions().canUseMethod(
-                                        CollectOptions::CollectMethod::HASH));
-
-    if (canUseHashAggregation && !opt->runOnlyRequiredRules(1)) {
-      if (collectNode->getOptions().shouldUseMethod(
-              CollectOptions::CollectMethod::HASH)) {
+                                        CollectOptions::CollectMethod::kHash));
+
+    if (canUseHashAggregation) {
+      bool preferHashCollect = collectNode->getOptions().shouldUseMethod(
+          CollectOptions::CollectMethod::kHash);
+
+      if (ExecutionNode const* loop = collectNode->getLoop();
+          loop != nullptr && loop->getLoop() == nullptr &&
+          (loop->getType() == EN::ENUMERATE_LIST ||
+           loop->getType() == EN::TRAVERSAL ||
+           loop->getType() == EN::ENUMERATE_PATHS ||
+           loop->getType() == EN::SHORTEST_PATH)) {
+        // if the COLLECT is contained inside a single loop, and the loop is an
+        // enumeration over an array (in contrast to an enumeration over a
+        // collection/view, then prefer the hashed collect variant. this is
+        // because the loop output is unlikely to be sorted in any way.
```

Comment:
I thought we had some optimization for sort `ENUMERATE_LIST` in case of optimization time known list

Reply:
The only optimization I know of that optimizes away sorts is the one that removes sorts for inputs which are produced by a sorted index. Maybe there exists also such optimization for inputs produced by a sorted view. 
But I am not aware of an optimization that checks the sortedness of an `ENUMERATE_LIST` input that is known at optimization time. We could write one though :sweat_smile: